adr/ADR-32.md

```diff
@@ -0,0 +1,179 @@
+\* | `js-delete-consumer(stream, name)` | `$JS.API.CONSUMER.DELETE.{stream}.{name}` | pub |
+\* | `js-delete-consumer(*, name)` | `$JS.API.CONSUMER.DELETE.*.{name}` | pub |
+\* | `js-consumer-names` | `$JS.API.CONSUMER.NAMES.*` | pub |
+\* | `js-consumer-names(stream)` | `$JS.API.CONSUMER.NAMES.{stream}` | pub |
+\* | `js-consumer-list` | `$JS.API.CONSUMER.LIST.*` | pub |
+\* | `js-consumer-list(stream)` | `$JS.API.CONSUMER.LIST.{stream}` | pub |
+\* | `js-consumer-info` | `$JS.API.CONSUMER.INFO.*.*` | pub |
+\* | `js-consumer-info(stream)` | `$JS.API.CONSUMER.INFO.{stream}.*` | pub |
+\* | `js-consumer-info(stream, name)` | `$JS.API.CONSUMER.INFO.{stream}.{name}` | pub |
+\* | `js-consumer-info(*, name)` | `$JS.API.CONSUMER.INFO.*.{name}` | pub |
+\* | `js-consumer-next-msg` | `$JS.API.CONSUMER.MSG.NEXT.*.*` | pub |
+\* | `js-consumer-next-msg(stream)` | `$JS.API.CONSUMER.MSG.NEXT.{stream}.*` | pub |
+\* | `js-consumer-next-msg(stream, name)` | `$JS.API.CONSUMER.MSG.NEXT.{stream}.{name}` | pub |
+\* | `js-consumer-next-msg(*, name)` | `$JS.API.CONSUMER.MSG.NEXT.*.{name}` | pub |
+\* | `js-consumer-ack-reply` | `$JS.ACK.*.*.>` | pub |
+\* | `js-consumer-ack-reply(stream)` | `$JS.ACK.{stream}.*.>` | pub |
+\* | `js-consumer-ack-reply(stream, name)` | `$JS.ACK.{stream}.{name}.>` | pub |
+\* | `js-consumer-ack-reply(*, name)` | `$JS.ACK.*.{name}.>` | pub |
+KeyValue | `kv-put` | `$KV.*.>` | pub
+\* | `kv-put(bucket)` | `$KV.{bucket}.>` | pub
+\* | `kv-put(bucket, key)` | `$KV.{bucket}.{key}` | pub
+\* | etc..
+
+*TODO: add remaining subjects, leader, peer, advisories, etc.*
+
+The second logical tier can be permission groups, often designed as *roles*, such as `js-stream-operator`. Parameters, such as stream or consumer names, will be transitively applied to the underlying permissions. The goal of roles is to abstract away a set of permissions for common use cases.
+
+This table shows a few examples.
+
+Name | Permissions | Notes
+---- | ----------- | -----
+`js-stream-operator` | `js-create-stream`, `js-update-stream`, `js-delete-stream` `js-purge-stream`, `js-stream-info`, `js-snapshot-stream`, `js-restore-stream`
+`js-stream-operator(name)` | `js-create-stream(name)`, `js-update-stream(name)`, etc...
+`js-stream-user(stream, subject)` | `publish(subject)`, `js-stream-info(stream)`, `js-stream-get-last-subject-msg(stream, subject)` | The `publish(subject)` is alternative to independently setting a pub-allow.
+
+
+The best way to model this in configuration is TBD, however, one approach could support `allow` and `deny` as top-level keys in the `permissions` map which would be used exclusively with these new logical permissions.
+
+For example:
+
+```
+accounts: {
+  APP: {
+    users: [
+      {
+        user: operator,
+        password: operator,
+        permissions: {
+          allow: [
+            "inbox(operator)",
+            "js-stream-operator(EVENTS)",
+          ]
```

Comment:
Maybe also need to cover how this maps to the imports and export syntax while we are at it, sub represent either a `stream` or `service`.
Like being able to define that reading info about consumer and streams is ok from the other account but not ok to create, something like: 

```hcl
  exports: [{stream: "foo"}, { jetstream: { streams: [ { name: EVENTS, export: [stream-info, consumer-info] }]
```

Reply:
Indeed. Fundamentally, these permissions can be thought of as functions that expand to one or more subjects `js-stream-info(EVENTS)` &rarr; `$JS.API.STREAM.INFO.EVENTS`. I thought about having these permissions able to be intermingled with standard subjects. If that is the case, the ambiguity of subject vs perm needs to be removed, so some kind of prefix character could help differentiate, such as `@` or whatever.

Then you could have something like:

```
exports: [
  {service: "@js-stream-info(EVENTS)"},
]
```

The implicit pub/sub information would really only apply in the user permissions context.